### PR TITLE
Dynamic anchor rate

### PR DESCRIPTION
**Pull request by systemv**
## Summary of changes

<Describe summary of major changes here>

## Report of required housekeeping

- [ ] Github issue OR spec proposal link
- [ ] Wrote tests
- [ ] Added a relevant changelog entry to CHANGELOG.md

----

## (FOR ADMIN) Before merging

- [ ] Added appropriate labels to PR
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" (coding standards)
- [ ] Confirm added tests are consistent with the intended behavior of changes
- [ ] Ensure all tests pass

### CODE REVIEW

File: contracts/overseer/src/contract.rs [hunk 1]
```diff
@@ -287,15 +311,97 @@ pub fn update_whitelist(
...
+    if !dynrate_state.prev_yield_reserve.is_zero()
+        && env.block.height > dynrate_state.last_executed_height + dynrate_config.dyn_rate_epoch
+    {
+        let blks = Uint256::from(env.block.height - dynrate_state.last_executed_height);
+
+        // yield reserve amt
+        let yield_reserve = Decimal256::from_uint256(interest_buffer);
+
+        // direction of rate change
+        let up_down = yield_reserve > dynrate_state.prev_yield_reserve;
```
Comment (alwin-peng) [inline contracts/overseer/src/contract.rs:353]: please rename this to something like `yr_went_up`, as something like `if up_down` provides no info on what the condition is

File: contracts/overseer/src/contract.rs [hunk 2]
```diff
@@ -287,15 +311,97 @@ pub fn update_whitelist(
...
+
+        // increase expectation pb
+        let increase_expectation_pb =
+            dynrate_config.dyn_rate_yr_increase_expectation / Decimal256::from_uint256(blks);
+
+        // consider increase expectation
+        yield_reserve_change_pb = if yield_reserve_change_pb > increase_expectation_pb {
+            yield_reserve_change_pb - increase_expectation_pb
+        } else {
+            yield_reserve_change_pb
```
Comment (alwin-peng) [inline contracts/overseer/src/contract.rs:373]: `yield_reserve_change_pb` means different things depending on `up_down`. here if the yield reserve went down, we will decrement the absolute change, when we actually want to increase it. so would suggest something like:

```rust
        yield_reserve_change_pb = if !up_down { 
            yield_reserve_change + increase_expectation_pb
        } else  if yield_reserve_change_pb > increase_expectation_pb {
            yield_reserve_change_pb - increase_expectation_pb
        } else {
            // sign switched here
            up_down = !up_down;
            increase_expectation_pb - yield_reserve_change_pb
        }

File: contracts/overseer/src/contract.rs [hunk 3]
```diff
@@ -287,15 +311,97 @@ pub fn update_whitelist(
...
+        };
+
+        // change exceeded rate threshold, need to update variable rate
+        // recalc values from confing to a per block
+        let year = Decimal256::from_str("4656810").unwrap();
```
Comment (alwin-peng) [inline contracts/overseer/src/contract.rs:378]: lets make this a constant and not use magic numbers

File: contracts/overseer/src/contract.rs [hunk 4]
```diff
@@ -308,16 +414,17 @@ pub fn execute_epoch_operations(deps: DepsMut, env: Env) -> Result<Response, Con
...
-    let mut messages: Vec<CosmosMsg> = vec![];
-    let mut interest_buffer = query_balance(
-        deps.as_ref(),
-        env.contract.address.clone(),
-        config.stable_denom.to_string(),
-    )?;
+    // # of blocks from the last executed height
+    let blocks = Uint256::from(env.block.height - state.last_executed_height);
+
+    let deposit_rate = update_rate(
```
Comment (alwin-peng) [inline contracts/overseer/src/contract.rs:423]: i think this should still be
```rust
    let deposit_rate =
        (effective_deposit_rate - Decimal256::one()) / Decimal256::from_uint256(blocks);
```
we're looking for the realized deposit rate, it shouldn't have anything to do with a dynamic rate calculation

Comment (alwin-peng) [reply contracts/overseer/src/contract.rs:423]: i think there might be a misunderstanding with what this function is supposed to do. currently what `execute_epoch_operations` does is:
1. perform ANC buybacks if the yield reserve increased
2. sends money from yield reserve to market contract if the realized interest rate is less than the threshold rate
3. notifies custody contracts to send staking rewards to the yield reserve

this change is supposed to add a fourth step (update the deposit rate). we need this step to be last, as we want step 2 to use the old deposit rate, and not retroactively apply the deposit rate changes. to make this more obvious, we should put `update_deposit_rate` in a separate function, that we run at the end of `execute_epoch_operations`

Comment (systemv) [reply contracts/overseer/src/contract.rs]: done

File: contracts/overseer/src/contract.rs (hunk 1 above)
Comment (systemv) [reply contracts/overseer/src/contract.rs]: done

File: contracts/overseer/src/contract.rs (hunk 2 above)
Comment (systemv) [reply contracts/overseer/src/contract.rs:370]: done

File: contracts/overseer/src/contract.rs (hunk 3 above)
Comment (systemv) [reply contracts/overseer/src/contract.rs]: good point, done

File: contracts/overseer/src/contract.rs [hunk 5]
```diff
@@ -287,15 +311,97 @@ pub fn update_whitelist(
...
+    // retrieve interest buffer
+    let mut messages: Vec<CosmosMsg> = vec![];
+    let mut interest_buffer = query_balance(
+        deps.as_ref(),
+        env.contract.address.clone(),
+        config.stable_denom.to_string(),
+    )?;
+
+    // check whether its time to re-evaluate rate
+    if !dynrate_state.prev_yield_reserve.is_zero()
```
[the remark this answers is not in the document]
Comment (systemv) [reply contracts/overseer/src/contract.rs:342]: done, with same naming conventions as 1.1

File: contracts/overseer/src/contract.rs [hunk 6]
```diff
@@ -514,6 +649,7 @@ pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
             deps.api.addr_validate(&borrower)?,
             block_time,
         )?),
+        QueryMsg::DynrateState {} => to_binary(&query_state(deps)?),
```
Comment (alwin-peng) [inline contracts/overseer/src/contract.rs:652]: should be `query_dynrate_state`

File: contracts/overseer/src/contract.rs [hunk 7]
```diff
@@ -51,6 +56,16 @@ pub fn instantiate(
         },
     )?;
 
+    store_dynrate_config(
```
Comment (alwin-peng) [inline contracts/overseer/src/contract.rs:59]: in practice this stuff will be migrated, and not instantiated. lets create a `migrate` hook, and have the migrate hook store dynrate config and state from the `MigrateMsg`.

Comment (alwin-peng) [reply contracts/overseer/src/contract.rs:59]: also, lets consider pruning state -- don't think `DynrateState.update_vector` or `DynrateState.rate_delta` are used anywhere anymore

Comment (systemv) [reply contracts/overseer/src/contract.rs:59]: done

File: contracts/overseer/src/contract.rs (hunk 6 above)
Comment (systemv) [reply contracts/overseer/src/contract.rs]: done

File: contracts/overseer/src/contract.rs [hunk 8]
```diff
@@ -287,6 +330,113 @@ pub fn update_whitelist(
...
+        let mut yield_reserve_change_pb =
+            yield_reserve_change / Decimal256::from_uint256(blocks_count);
+
+        // increase expectation pb
+        let increase_expectation_pb = dynrate_config.dyn_rate_yr_increase_expectation
+            / Decimal256::from_uint256(blocks_count);
+
+        // consider increase expectation
+        yield_reserve_change_pb = if !yr_went_up {
+            yield_reserve_change + increase_expectation_pb
```
Comment (jjeogu) [inline contracts/overseer/src/contract.rs:387]: when the yield reserve does not go up, the yield_reserve_change_pb is set by using the total yield reserve change?

given the difference in units of the yield_reserve_change and yield_reserve_change_pb, i would guess the target deposit rate would get reduced by the dynrate_maxchange_pb very often.

On the other hand, when the yield reserve goes up, the yield_reserve_change_pb is set in "per-block" units.

is this asymmetric rate adjustment intended?

File: contracts/overseer/src/contract.rs [hunk 9]
```diff
@@ -287,6 +330,113 @@ pub fn update_whitelist(
...
+
+        // increase expectation pb
+        let increase_expectation_pb = dynrate_config.dyn_rate_yr_increase_expectation
+            / Decimal256::from_uint256(blocks_count);
+
+        // consider increase expectation
+        yield_reserve_change_pb = if !yr_went_up {
+            yield_reserve_change_pb + increase_expectation_pb
+        } else if yield_reserve_change_pb > increase_expectation_pb {
+            yield_reserve_change_pb - increase_expectation_pb
```
[the remark this answers is not in the document]
Comment (jjeogu) [reply contracts/overseer/src/contract.rs:389]: ah got it, the 3rd case logic tripped me up

thanks for the explanation!

File: contracts/overseer/src/contract.rs [hunk 10]
```diff
@@ -287,6 +330,114 @@ pub fn update_whitelist(
...
+        } else {
+            // sign switched here
+            yr_went_up = !yr_went_up;
+            increase_expectation_pb - yield_reserve_change_pb
+        };
+
+        // change exceeded rate threshold, need to update variable rate
+        // recalc values from confing to a per block
+        let _year = Decimal256::from_uint256(BLOCKS_PER_YEAR);
```
Comment (kevko2020) [inline contracts/overseer/src/contract.rs:398]: is this used anywhere?

Comment (systemv) [reply contracts/overseer/src/contract.rs:398]: Not used anymore/will be removed, as it was decided that both configuration options should be per epoch. Thank you!

File: contracts/overseer/src/contract.rs [hunk 11]
```diff
@@ -287,6 +378,90 @@ pub fn update_whitelist(
...
+        yield_reserve_change = if !yr_went_up {
+            yield_reserve_change + increase_expectation
+        } else if yield_reserve_change > increase_expectation {
+            yield_reserve_change - increase_expectation
+        } else {
+            yr_went_up = !yr_went_up;
+            increase_expectation - yield_reserve_change
+        };
+
+        yield_reserve_change = min(yield_reserve_change, dynrate_config.dyn_rate_maxchange);
```
Comment (ekryski) [inline contracts/overseer/src/contract.rs:430]: What happens if LFG tops up the YR? I suspect we'll see the rates drop a few steps first:

* April: 18%
* May: 16.5%

So adding another 3% back after dropping 3% will keep the rate where it's at. But if TFL tops up before hitting 16.5%, couldn't the `new_rate` theoretically get to 21%?

Maybe you want to guard against a rapid increase in YR? Apologies if I missed it somewhere.

File: contracts/overseer/src/contract.rs [hunk 12]
```diff
@@ -287,6 +378,90 @@ pub fn update_whitelist(
...
+        let mut new_rate = if yr_went_up {
+            current_rate + yield_reserve_change
+        } else if current_rate > yield_reserve_change {
+            current_rate - yield_reserve_change
+        } else {
+            Decimal256::zero()
+        };
+
+        // convert from yearly rate to block rate
+        new_rate = new_rate / blocks_per_year;
```
Comment (ekryski) [inline contracts/overseer/src/contract.rs:441]: Might be cleaner to just give this a name like `new_per_block_rate` so its easier to follow.

File: contracts/overseer/src/contract.rs [hunk 13]
```diff
@@ -482,6 +671,26 @@ pub fn update_epoch_state(
         ]))
 }
 
+pub fn fund_reserve(deps: DepsMut, info: MessageInfo) -> Result<Response, ContractError> {
+    let sent_uusd = match info.funds.iter().find(|x| x.denom == "uusd") {
```
Comment (jjeogu) [inline contracts/overseer/src/contract.rs:675]: any reason not to use the config struct's stable_denom field?

File: contracts/overseer/src/contract.rs (hunk 11 above)
Comment (kevko2020) [reply contracts/overseer/src/contract.rs:438]: Should we have some limit to make sure we don't go over 20%? Or does this dynamic change means it's possible to keep increasing as long as the reserve is somehow growing?

**Pull request merged**